app/controllers/application_controller.rb

```diff
@@ -5,6 +5,144 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, 'totally_secure_key'
   end
 
-end
\ No newline at end of file
+  get "/" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :index
+    end
+  end
+
+  get "/signup" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :signup
+    end
+  end
+
+  post "/signup" do
+    user = User.new(params)
+    if user.save
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/signup"
+    end
+  end
+
+  get "/login" do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :login
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+  get "/logout" do
+    session.clear
+    redirect "/login"
+  end
+
+  get "/tweets" do
+    as_logged_in do
+      @tweets = Tweet.all
+      erb :'tweets/index'
+    end
+  end
+
+  post "/tweets" do
+    as_logged_in do |user|
+      tweet = Tweet.new(content: params[:content], user: user)
+      if tweet.save
+        redirect "/tweets"
+      else
+        redirect "/tweets/new"
+      end
+    end
+  end
+
+  get "/tweets/new" do
+    as_logged_in do
+      erb :'tweets/new'
+    end
+  end
+
+  get "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show'
+    end
+  end
+
+  patch "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      if user==@tweet.user && @tweet.update(content: params[:content])
+        redirect "/tweets/#{params[:id]}"
+      else
+        redirect "/tweets/#{params[:id]}/edit"
+      end
+    end
+  end
+
+  delete "/tweets/:id" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      if user==@tweet.user && @tweet.delete
+        redirect "/tweets"
+      else
+        redirect "/tweets/#{params[:id]}"
+      end
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    as_logged_in do |user|
+      @tweet = Tweet.find(params[:id])
+      if user==@tweet.user
+        erb :'tweets/edit'
+      else
+        redirect "/tweets/#{params[:id]}"
+      end
+    end
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  helpers do
+    def logged_in?
+      session[:user_id]!=nil
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```